Better view relations

calebmer opened the pull request:
So here is my situation, this is what most of the tables that I use with PostgREST look like.

```sql
CREATE VIEW api.person AS
  SELECT id AS "id", given_name AS "givenName", family_name AS "familyName",
         image AS "image",
         comment_group_id AS "commentGroupId", api.comment_count(comment_group_id) AS "commentCount",
         star_group_id AS "starGroupId", api.star_count(star_group_id) AS "starCount",
         created_at AS "createdAt"
    FROM devtips.person;

CREATE VIEW api.post AS
  SELECT id AS "id", headline AS "headline", author_id AS "authorId",
        post_topic_id AS "postTopicId", is_based_on_url AS "isBasedOnUrl",
        comment_group_id AS "commentGroupId", api.comment_count(comment_group_id) AS "commentCount",
        star_group_id AS "starGroupId", api.star_count(star_group_id) AS "starCount",
        created_at AS "createdAt", text AS "text"
    FROM devtips.post;
```

A view that refers to other views in the schema and renames all of the columns. I believe this is the best design as it abstracts away the internal implementation and allows me to add extra columns (such as `starCount`). I want PostgREST to work in such a way that it detects a relation between `authorId` (a renamed column) and `api.person` (a view and not the source relation). Thus, this PR happens.

It accomplishes this by instead of selecting view columns in the `all*` functions, it selects column synonyms (which it detects by regular expression on `information_schema.views.view_definition`) and then recreates all of the primary keys/relations that way. The PR also adds code which looks at all the relations, and if the relation has foreign columns in a schema PostgREST doesn't use, and there are synonymous columns in a the schema PostgREST does use, the code will swap out those two column sets.

Also in this PR is a large refactor of `DbStructure` which makes the object *much* easier to work with, and better for future development. The changes made in `DbStructure` can be seen in both `0214126` and `860742f`.

These were the changes I warned about in #342. @ruslantalpa I feel it useful for `DbStructure` to have a `tables` field, therefore my merge strategy was to only query for accessibility on every request and use that to filter the pre-queried tables. Thoughts?

(**PS**: Made it! #289 @bitemyapp Thanks to everyone who answered my newbie Haskell questions 😊)

calebmer commented:
This one is going to need some critical review. Oh yeah, and one test is failing. I have no clue why, while you guys are reviewing maybe you could try spotting the source?

I narrowed it down so that I *think* the faulty code is having some adverse effect [here](https://github.com/calebmer/postgrest/blob/feature/view-relations/src/PostgREST/QueryBuilder.hs#L47-L69).

ruslantalpa commented:
Can you provide  the create statements of a few tables/views to that i can run your queries and see what relations are detected (to better understand how the code works).
Are you saying this PR will detect relations like "author_id AS myCustomNameId" or just  "author_id AS authorId" ?

If it's just the second case then i would be against it since it's a very limited case (based on your specific preference for camelCase in javascript) and the detection method seems a bit fragile (regexp).  I would rather provide the api client the ability to rename each column if he choses to anything he likes (while internally postgrest works with real column names). If the code can detect "id AS myCustomName" then i am for it but my feel is that this PR is driven by the desire to have "myColumnName" in javascript as opposed to "my_column_name" that you have in the db (and i would advice against going this route, you are creating a lot of views just for the sake of renaming columns). I would say that this particular problem/need is better addressed by providing the ability for the api client to rename columns in output on the fly (with much smaller code footprint). Beside allowing for conversion to camelCase, it also allow this:
When including a parent item, currently it will be named the way the table is names (clients). It would be nice to be able to shape the output so that the key is "client" (singular). This ability to rename columns is also needed in order to be isomorphic to graphql.

I have not really reviewed the code carefully so take the above with a grain of salt. I'll take another pass when you post a few tables/views create statements

ruslantalpa commented:
I forgot why i added relation detection for views (probably because 9.4 did not have RLS and they were useful for that) but i fear that when joining two views based on some FK that lives inside the original table, the index on the FK in the original table might not be used thus leading to some terrible performance. This needs investigation but if it's true (indexes from original tables not being used for joins) then i wonder if this feature (detecting relations for views) is not encouraging bad db design (and performance issues will be blamed on postgrest)

calebmer commented:
The way this is designed it will work for any name, so `author_id AS myCustomNameId` will be detected.  For me, camelCase has always been a nice side effect that I choose to opt-in on. Its a style decision that I would like PostgREST to support, but I wouldn't die without it.

The primary motivation is for something like this: `author_id AS author`. The next thing I am working on is allowing for the `select` parameter to detect relations in columns, for example: `?select=id,name,author_id(*)`. This PR allows for that to become `?select=id,name,author(*)`. In addition this PR is needed because the current system expects the names to be the same, which is faulty and can cause problems.

As for the fragility of the regex: I am confident it will work 90% of the time and that final 10% the user is doing something which should not be supported and the regex drops it from consideration. [This](https://github.com/begriffs/postgrest/pull/353/files#diff-43dc87b0189a597685a9a5eb8cec3f41R330) is the line which gives me the confidence to say that.

We should also consider what is good Postgres database design, and what is good PostgREST database design. I believe views are the best PostgREST design as they allow the developer to hide underlying workings of the database, they allow for data security (columns users shouldn't edit), and they allow for dynamic columns (as you can see I have a `api.star_count(star_group_id)` column).

calebmer commented:
@ruslantalpa [here](https://gist.github.com/calebmer/ed4e6bd1cb257a8c86a7) is a subset of a schema I'm optimizing for.

bitemyapp commented:
@calebmer :heart:

begriffs commented:
If it helps debugging, I noticed there are sql errors from the fixture on CI:

```
psql:test/fixtures/schema.sql:454: ERROR:  relation "articles_id_seq" does not exist
LINE 1: SELECT pg_catalog.setval('articles_id_seq', 1, false);
                                 ^
psql:test/fixtures/schema.sql:518: ERROR:  multiple primary keys for table "articles" are not allowed
psql:test/fixtures/schema.sql:293: ERROR:  type "jwt_claims" already exists
psql:test/fixtures/schema.sql:454: ERROR:  relation "articles_id_seq" does not exist
LINE 1: SELECT pg_catalog.setval('articles_id_seq', 1, false);
                                 ^
psql:test/fixtures/schema.sql:518: ERROR:  multiple primary keys for table "articles" are not allowed
```

ruslantalpa commented:
On the first pass the PR is ok, the only suggestions i have is renaming a few functions to make it explicit what they are doing.

PS: on the author_id(*) thing, please don't start building it before discussing, first feeling is "it's a bad direction". I see the problem with author_id editor_id but i am not completely sure it's such a common situation (maybe others can provide more examples) and even more i do not like the author_id(*) interface because it decouples the words used in "select" and the keys that appear in response. Let's discuss this.

calebmer commented:
Ok, making the changes. We can discuss the `author_id(*)` approach in #354 as it is (mostly) unrelated to this PR.

ruslantalpa commented:
To make a note here about what was discussed the chat.
The fromJust function should be eliminated from the code

Ok, here's the inventory:
raiseRelations: Ok because of isJust check in guard
columnFromRow
relationFromRow
pkFromRow
allSynonyms

calebmer commented:
The places `fromJust` is, I'm confident are completely fine because of the implicit promise the database makes that every column has a table (etc.). The usage should be looked at anyway and if anyone has a recommendation on how else to handle it, that would be great.

calebmer commented:
Ok, fixed the tests and removed the `fromJust` assumptions. Now if something is not returned using `find` (which theoretically shouldn't happen) it will just be dropped from the result set.

calebmer commented:
Ok, used some new Applicative stuffs to make the code more readable.

ruslantalpa commented on src/PostgREST/DbStructure.hs:
```diff
@@ -0,0 +1,354 @@
+  tabs <- allTables
+  cols <- allColumns tabs
+  syns <- allSynonyms cols
+  rels <- allRelations tabs cols
```
maybe rename from
rels to childRelations
then you can have the line below look like
```
rels = (addManyToManyRelations . raiseRelations schema syns . addParentRelations . addSynonymousRelations syns) childRelations
```

calebmer replied:
Doing that kinda breaks the nice 4 character length going on there 😊

I can do it, but maybe we should consider renaming the other ones too?

begriffs commented on src/PostgREST/MainTest.hs:
```diff
@@ -8,7 +8,7 @@ import           PostgREST.Config                     (AppConfig (..),
                                                        readOptions)
```
Actually I don't know what this file is. I deleted it locally and was able to build the binary and run the tests, so I'd say get rid of it.

calebmer replied:
Do you know what `mock.hs` is?

diogob replied:
I was wondering the same thing!

ruslantalpa replied:
that was me, added it by mistake, it can be deleted

begriffs commented on src/PostgREST/DbStructure.hs:
```diff
@@ -0,0 +1,354 @@
+allRelations :: [Table] -> [Column] -> H.Tx P.Postgres s [Relation]
+allRelations tabs cols = do
+  rels <- H.listEx $ [H.stmt|
+    SELECT ns1.nspname AS table_schema,
```
The queries in the code including this one are ingenious and a marvel, but... really intense! Feels like we're releasing Zalgo.

![sql-zalgo](https://cloud.githubusercontent.com/assets/911911/11161216/6e68f42e-8a2b-11e5-9acb-93b5dfeda911.png)

Some kind of comment on the function would be helpful, like showing an example of what results are returned by the sql inside. I know it'll go through `relationFromRow` and be turned into a `Relation` type, but the comment would be helpful.

calebmer replied:
I did not actually write that SQL query, that was in the original PgStructure.hs so I can't help there 😐

begriffs replied:
Sorry, got confused about what was added vs rearranged.

begriffs commented on src/PostgREST/Types.hs:
```diff
@@ -6,65 +6,68 @@ import Data.Aeson
 data Table = Table {
-  tableSchema :: Text
-, tableName :: Text
+  tableSchema     :: Schema
```
Nice consolidation!

begriffs commented on src/PostgREST/Types.hs:
```diff
@@ -6,65 +6,68 @@ import Data.Aeson
+, dbPrimaryKeys :: [PrimaryKey]
+} deriving (Show, Eq)
 
+type Schema = Text
```
Love it. Makes the signatures easier to read all around the code.

diogob commented on src/PostgREST/Config.hs:
```diff
@@ -28,7 +28,7 @@ import           Data.Text                   (strip)
 import           Network.Wai
 import           Network.Wai.Middleware.Cors (CorsResourcePolicy (..))
-import           Options.Applicative         hiding (columns)
+import           Options.Applicative
```
I'd suggest we avoid importing any module without either qualifying or using an explicit list of what is being imported. Of course there are exceptions to the rule, but most times we need just a couple of functions.
Explicit importing reduces name clashes, makes the code easier to refactor and readable for people not familiar with the used libraries.

calebmer replied:
I agree, that should probably be it's own PR though.

diogob commented on src/PostgREST/App.hs:
```diff
@@ -289,7 +290,7 @@ convertJson v = (,) <$> (header <$> normalized) <*> (vals <$> normalized)
         _ -> Left invalidMsg
 
-augumentRequestWithJoin :: Text ->  [Relation] ->  ApiRequest -> Either Text ApiRequest
+augumentRequestWithJoin :: Schema ->  [Relation] ->  ApiRequest -> Either Text ApiRequest
```
Nice touch! :D

begriffs commented on src/PostgREST/App.hs:
```diff
@@ -147,7 +147,7 @@ app dbstructure conf reqBody req =
 
     ([], _) -> do
-      body <- encode <$> tables (cs schema)
+      body <- encode <$> accessibleTables (filter ((== cs schema) . tableSchema) allTabs)
```
What if `dbTables` was already filtered to be just the accessible ones? Or are all the tables needed in order to do the relation lookups?

calebmer replied:
`dbTables` is initialized with all tables in the database except the ones in the `pg_catalog` and `information_schema` schemas. This is how I merged #342, we prefetch all the tables and then filter every request which ones the user actually has access to (vs selecting only the ones the user has access to from the outset).

begriffs commented on src/PostgREST/App.hs:
```diff
@@ -332,10 +333,10 @@ addFilter (path, flt) (Node rn forest) =
-  | mt == t = Just $ r {relTable=sourceSubqueryName}
-  | mt == ft = Just $ r {relFTable=sourceSubqueryName}
-  | Just mt == rt = Just $ r {relLTable=Just sourceSubqueryName}
+toSourceRelation mt r@(Relation t _ ft _ _ rt _ _)
```
This function confuses me, and leads me into other confusing code as well. I went to see how it is used, and found it in `parseRequest` and wow that did not clear things up.

Maybe it would help to not pattern match into tiny variable names. I don't know what `mt` means, or the difference between `relFTable` and `relLTable` inside `Relation`. With more digging through I'll be able to piece it together but these names aren't helping.

calebmer replied:
Wish I could help here, but when I switched all the types I just went through all the things that were throwing compilation errors and fixing them. If you think that's confusing just take a look at [this](https://github.com/calebmer/postgrest/commit/acb8f8a15491352fcad6beda7b1df6f9515822fc) fix for the failing test.

begriffs commented on src/PostgREST/DbStructure.hs:
```diff
@@ -0,0 +1,354 @@
+  let isAccessible table = isJust $ find (\(s,n) -> tableSchema table == s && tableName table == n) accessible
+  return $ filter isAccessible allTabs
+
+synonymousColumns :: [(Column,Column)] -> [Column] -> [[Column]]
```
I haven't wrapped my head around what this does, but when I see `[(Column,Column)]` it makes me wonder if `Map Column Column` would be helpful from [containers](https://hackage.haskell.org/package/containers-0.5.6.3/docs/Data-Map-Strict.html).

ruslantalpa replied:
no, (,) is perfect here because what (column,column) means is that they are synonyms, like a column from a table and a (renamed) column from a view based on that table

calebmer replied:
Is there a map (or other data type) which can find columns in both directions?

begriffs commented:
Another option is to merge this as it is and try to get the whole v3 thing back into master. Then once back on master do some major cleanup (module reorganization, improve names, add comments) etc before proceeding with new features. Let me know what you'd like to do, if you'd rather address any of the code review items before merging that's fine by me as well.

ruslantalpa commented:
i am in favor of merging this, then merging v3 to master then refactoring
